src/featbin/generate-random-cmn-offsets.cc

```diff
@@ -0,0 +1,289 @@
+    po.Register("cmn-offset-scale", &cmn_offset_scale, "offset scale used to scale the covariance matrix.");
+    po.Register("preserve-total-covariance", &preserve_total_covariance, "If true, the total covariance for random"
+                " offset of spks are preserved.");
+    po.Register("weights", &weights_rspecifier, "rspecifier for a vector of floats "
+                "for each utterance, that's a per-frame weight.");
+    po.Register("zero-mean-offsets", &zero_mean, 
+                "If true, the random offsets generated using zero mean distribution.");
+    po.Register("cov-self-check", &check_offset_spk_cov, 
+                "If true, the spk covariance for new shifted speakers are computed"
+                " and eigenvalues for orig cov^-1 * shifted-cov is printed"
+                " for self-check.");  
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 3) {
+      po.PrintUsage();
+      exit(1);
+    }
+   
+    std::string spk2utt_rspecifier = po.GetArg(1),
+     feat_rspecifier = po.GetArg(2),
+      cmn_wspecifier = po.GetArg(3);
+    DoubleMatrixWriter cmn_writer(cmn_wspecifier); 
+   
+    srand(srand_seed);
+    int32 num_done = 0, num_err = 0, feat_dim = 0;
+
+    SequentialTokenVectorReader spk2utt_reader(spk2utt_rspecifier);
+    RandomAccessBaseFloatMatrixReader feat_reader(feat_rspecifier); 
+
+    RandomAccessBaseFloatVectorReader weights_reader(weights_rspecifier);
+
+    Matrix<double> global_stats,
+      spk_cov;
+    bool is_global_init = false;
+    int32 num_spks = 0;
+    Vector<double> spk_mean_sum;
+    std::map<std::string, Vector<double> > spk_means;
+    std::map<std::string, std::vector<std::string> > spk_uttlist;
+    for (; !spk2utt_reader.Done(); spk2utt_reader.Next()) {
+      std::string spk = spk2utt_reader.Key(); 
+      const std::vector<std::string> &uttlist = spk2utt_reader.Value();
+      spk_uttlist[spk] = uttlist;
+      bool is_init = false;
+      Matrix<double> stats;
+      for (size_t i = 0; i < uttlist.size(); i++) {
+        std::string utt = uttlist[i];
+        if (!feat_reader.HasKey(utt)) {
+            KALDI_WARN << "Did not find features for utterance " << utt;
+            num_err++;
+            continue;
+        }
+        const Matrix<BaseFloat> &feats = feat_reader.Value(utt);
+        // Initialize local stat to accumulate stats per speaker
+        if (!is_init) {
+          InitCmvnStats(feats.NumCols(), &stats);
+          feat_dim = feats.NumCols();
+          is_init = true;
+        }
+
+       
+        if (!AccCmvnStatsWrapper(utt, feats, &weights_reader, &stats)) 
+          num_err++;
+        else
+          num_done++;
+      }
+      Vector<double> spk_mean(stats.Row(0).Range(0, feat_dim));
+      double count = stats(0, feat_dim);
+      spk_mean.Scale(1.0/count);
+      spk_means[spk] = spk_mean;
+      // Initialize global stats to accumulate stats for all data.
+      if (!is_global_init) {
```

Comment:
Can use `if (spk_cov.NumRows() == 0)` or `global_stats.Dim() == 0` or some condition like that.